docs/reconciler-proposal/kris-1/interface.go

```diff
@@ -0,0 +1,62 @@
+// Copyright © 2017 The Kubicorn Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud
+
+import (
+	"github.com/kris-nova/kubicorn/apis/cluster"
+)
+
+// Model will model resources to an API representation of infrastructure
+type Model interface {
+
+	// Get will return the map of ordered resources against a cloud
+	Get() map[int]Resource
+}
+
+// Reconciler will create and destroy infrastructure based on an intended state. A Reconciler will
+// also audit the expected and actual state.
+type Reconciler interface {
+
+	//SetModel will set the model to use for the cloud reconciler
+	SetModel(model Model)
+
+	// GetActual will audit a cloud and return the API representation of the current resources in the cloud
+	GetActual(known *cluster.Cluster) (*cluster.Cluster, error)
+
+	// GetExpected will audit a state store and return the API representation of the intended resources in the cloud
+	GetExpected(known *cluster.Cluster) (*cluster.Cluster, error)
+
+	// Reconcile will take an actual and expected API representation and attempt to ensure the intended state
+	Reconcile(actual, expected *cluster.Cluster) (*cluster.Cluster, error)
+
+	// Destroy will take an actual API representation and destroy the resources in the cloud
```

Comment:
This is a great point.

With this approach `Destroy` would take an `actual` which is what is running in the cloud and destroy the resource that are modeled out of the API. I don't see a case where we would ever have a skew unless the calling code passed in a representation that was different than what an `actual` would be at runtime. 

In other words, `Destroy` takes an actual so we know we are deleting resources that actually exist. If a user passed in an `expected` API (or any API for that matter) they would inherently be using `Destroy` wrong.

It would be possible to do an asserting at runtime during the beginning of a `Destroy` but that would double our efforts of querying the cloud as we would expect a user to call `GetActual` right before we call the same method to assert they are the same.

What are your thoughts here? Does this seem sane?